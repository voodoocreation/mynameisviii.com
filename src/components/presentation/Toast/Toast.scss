@import "../../../scss/variables", "../../../scss/mixins";

.Toast {
  opacity: 1;
  transition: opacity 200ms ease-out;

  &--content {
    @include panel-1();

    display: inline-flex;
    font-size: 14px;
    margin-top: $spacing-half;
    padding: $spacing;
    border: $text-color 1px solid;
    border-radius: $radius;
  }

  &--message {
    display: flex;
    flex-grow: 1;
    align-items: center;

    button {
      margin-left: $spacing;
    }
  }

  &--closeButton {
    margin: (-$spacing) (-$spacing) (-$spacing) 0;
    padding: $spacing;
  }

  &.isHidden {
    opacity: 0;
  }

  @include screen-below($screen-xs-max) {
    &--content {
      display: flex;
    }

    &--message {
      flex-direction: column;

      button {
        width: 100%;
        margin-left: 0;
        margin-top: $spacing-half;
      }
    }
  }
}
