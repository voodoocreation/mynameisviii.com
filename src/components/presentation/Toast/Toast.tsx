import cn from "classnames";
import * as React from "react";
import { MdClose } from "react-icons/lib/md";
import { InjectedIntl, injectIntl } from "react-intl";

interface IProps {
  autoDismissDelay: number;
  className?: string;
  hasAutoDismiss: boolean;
  intl: InjectedIntl;
  isVisible?: boolean;
}

interface IState {
  isRendered: boolean;
  isToggled: boolean;
  isVisible: boolean;
}

class Toast extends React.Component<IProps, IState> {
  public static defaultProps = {
    autoDismissDelay: 5000,
    hasAutoDismiss: true,
    isVisible: true
  };

  private dismissTimeOut: number | undefined = undefined;
  private visibilityTimeout: number | undefined = undefined;

  constructor(props: IProps) {
    super(props);

    this.state = {
      isRendered: !!props.isVisible,
      isToggled: !!props.isVisible,
      isVisible: false
    };
  }

  public componentDidMount() {
    this.toggleVisibility(!!this.props.isVisible);
  }

  public componentWillReceiveProps(nextProps: IProps) {
    if (nextProps.isVisible !== this.state.isToggled) {
      this.toggleVisibility(!!nextProps.isVisible);
    }
  }

  public render() {
    const { children, className } = this.props;
    const { formatMessage } = this.props.intl;
    const { isRendered, isVisible } = this.state;

    return !isRendered ? null : (
      <div className={cn("Toast", className, { isHidden: !isVisible })}>
        <div className="Toast-content">
          <div className="Toast-message">{children}</div>

          <button
            className="Toast-closeButton"
            onClick={this.onClose}
            title={formatMessage({ id: "CLOSE" })}
          >
            <MdClose />
          </button>
        </div>
      </div>
    );
  }

  private toggleVisibility = (isVisible: boolean) => {
    this.setState({
      isToggled: isVisible,
      [isVisible ? "isRendered" : "isVisible"]: isVisible
    } as any);

    if (this.visibilityTimeout) {
      window.clearTimeout(this.visibilityTimeout);
    }

    this.visibilityTimeout = window.setTimeout(() => {
      this.setState({
        [isVisible ? "isVisible" : "isRendered"]: isVisible
      } as any);

      if (isVisible && this.props.hasAutoDismiss) {
        if (this.dismissTimeOut) {
          window.clearTimeout(this.dismissTimeOut);
        }

        this.dismissTimeOut = window.setTimeout(
          this.toggleVisibility.bind(this, false),
          this.props.autoDismissDelay
        );
      }
    }, 200);
  };

  private onClose = () => {
    this.toggleVisibility(false);
  };
}

export default injectIntl<any>(Toast);
