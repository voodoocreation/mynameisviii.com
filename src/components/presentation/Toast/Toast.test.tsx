import { mount, render } from "enzyme";
import * as React from "react";

import Toast from "./Toast";

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    className: "TestToast",
    ...fromTestProps
  };

  return {
    actual: fn(<Toast {...props}>Test</Toast>),
    props
  };
};

describe("[presentation] <Toast />", () => {
  beforeEach(() => {
    window.setTimeout = jest.fn((fn: any) => {
      fn();
      return 1;
    });
    window.clearTimeout = jest.fn();
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it("renders correctly", () => {
    const { actual } = setup(render);

    expect(actual).toMatchSnapshot();
  });

  it("closes when close button is clicked, then reopens when new props are passed", () => {
    const { actual } = setup(mount, { isVisible: true, hasAutoDismiss: false });

    expect(actual.find(".Toast")).toHaveLength(1);

    actual.find(".Toast-closeButton").simulate("click");
    expect(actual.find(".Toast")).toHaveLength(0);

    actual.setProps({ isVisible: true });
    expect(actual.find(".Toast")).toHaveLength(1);
  });

  it("dismisses automatically when hasAutoDismiss=true", () => {
    const { actual } = setup(mount, { isVisible: true });

    actual.setProps({ isVisible: true });

    expect(window.setTimeout).toHaveBeenCalledTimes(6);
    expect(window.clearTimeout).toHaveBeenCalledTimes(3);
  });
});
