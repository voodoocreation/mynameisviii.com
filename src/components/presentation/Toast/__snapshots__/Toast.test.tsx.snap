// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Toast /> renders correctly 1`] = `
<div
  class="Toast TestToast isHidden"
>
  <div
    class="Toast-content"
  >
    <div
      class="Toast-message"
    >
      Test
    </div>
    <button
      class="Toast-closeButton"
      title="Close"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style="vertical-align:middle"
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m31.6 10.7l-9.3 9.3 9.3 9.3-2.3 2.3-9.3-9.3-9.3 9.3-2.3-2.3 9.3-9.3-9.3-9.3 2.3-2.3 9.3 9.3 9.3-9.3z"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
